src/geff/spatial_graph/io.py

```diff
@@ -36,51 +38,83 @@ def write(
 
     # write node attributes
     for name in graph.node_attr_dtypes.keys():
-        group[f"nodes/attrs/{name}"] = graph.node_attrs[nodes].__getattr__(name)
+        group[f"nodes/attrs/{name}/values"] = graph.node_attrs[nodes].__getattr__(name)
 
     # write edges
     group["edges/ids"] = graph.edges
 
     # write edge attributes
     for name in graph.edge_attr_dtypes.keys():
-        group[f"edges/attrs/{name}"] = graph.edge_attrs[edges].__getattr__(name)
+        group[f"edges/attrs/{name}/values"] = graph.edge_attrs[edges].__getattr__(name)
+
+
+def _read_attr_values(
+    geff_group: zarr.Group, attr_type: str, default_attr_values: dict[str, float] | None
+) -> tuple[dict[str, str], dict[str, np.ndarray]]:
+    """Read the values for a set of node or edge attributes, including checking for
+    missing values and either throwing an error or using the provided default.
+
+    Args:
+        geff_group (zarr.Group): The zarr group containing the geff_version in .zattrs
+        attr_type (str): "nodes" or "edges"
+        default_attr_values (dict[str, float]): A dictionary mapping attribute names
+            to default values to use if the attribute is missing
+
+    Raises:
+        ValueError: If an attribute has missing values and no default value for it is
+            provided
+
+    Returns:
+        tuple[dict[str, str], dict[str, np.ndarray]]: A mapping from attribute names to
+            dtypes, and a mapping rom attribute names to values as numpy arrays
+    """
+    attr_dtypes = {}
+    attrs = {}
+    assert attr_type in ["nodes", "edges"]
+    for name in geff_group[f"{attr_type}/attrs"]:
+        attrs_group = geff_group[f"{attr_type}/attrs/{name}"]
+        values = attrs_group["values"][:]
+        if "missing" in attrs_group.array_keys():
+            missing = attrs_group["missing"][:]
+            if np.any(missing):
+                if default_attr_values is None or name not in default_attr_values:
+                    raise ValueError(
+                        f"{attr_type.capitalize()} attribute {name} has missing values, "
+                        "must provide default value to read with spatial graph"
+                    )
+                values[missing] = default_attr_values[name]
+        dtype = values.dtype
+        size = values.shape
+        dtype_str = str(dtype) + ("" if len(size) == 1 else f"[{size[1]}]")
+        attr_dtypes[name] = dtype_str
+        attrs[name] = values
+    return attr_dtypes, attrs
 
 
-def read(path):
+def read(
+    path: Path | str,
+    validate: bool = True,
+    default_node_attr_values: dict[str, float] | None = None,
+    default_edge_attr_values: dict[str, float] | None = None,
+):
```

Comment:
Docstring please 🙏

Reply:
Not adding since SG is not in main anymore